### DISCUSSION TITLE
Allow slashes in filenames

### DISCUSSION THREAD

**Discussion by PatrikKozak** [payloadcms/payload#2911] [Feature Requests & Ideas]
Being that most blob storage "folders" are really just files with slashes in them, might consider ability to upload files that contain slashes in their filenames so that they are uploaded to "folders".

The problem is, we use the `sanitize-filename` package which automatically removes slashes from filenames - so we should figure out a way to either allow slashes specifically or have an option to opt out of filename sanitization.

**Comment by mrobst** [comment]
100% - I was just about to raise a discussion about this having been fighting for a couple of days to get Payload to behave the way I want it to....... I think this might be a fairly easy answer. 

Example use case: 

- Multiple users with accounts on payload uploading images for processing. 
- Image storage on any major cloud service. 
- Users want to get their image back (download) later with the same filename so they can match the files up again. 
- Image uploads for the same user should overwrite existing ones

Previous solution (not on Payload):
- Prefix all uploads for a user with their account id (unique) and "/" which creates a virtual folder in cloud storage
- 12345/image1.jpg
- 34567/image1.jpg
- Above example is for two different user accounts - all works fine

Payload currently doesn't allow "/" in the filename and expects all filenames to be unique across the whole application which is enforced by a unique index on filename in MongoDB which is automatically added on startup. 

There is a way to manipulate the prefix on the cloud storage plugin to be dynamically named which means that you can write files to folders in cloud storage via upload field but as this is the prefix it doesn't change the filename. Payload checks for the file existence when image1.jpg is uploaded for the second user and doesn't take the prefix into account. The file is saved into the second folder but when MongoDB is updated it throws a unique error on the filename as it sees 2 "image1.jpg" filenames. 

Allowing "/" in the filename would mean that the full name is stored in MongoDB and the unique index would still be valid. 

At the moment I have to avoid using the upload enabled collections on Payload and just store URLs to storage as text which isn't ideal :-(

@PatrikKozak did you find any workaround? Can I influence the MongoDB index to use prefix and filename fields together? 

References: 
Hacking the prefix for folders...
https://discord.com/channels/967097582721572934/1184466437331488839/1204564297515012118

Uniqueness & santization (I want to avoid this!)
https://github.com/payloadcms/payload/discussions/1153#discussioncomment-3702169

Folder view- this might be part of the solution
https://github.com/payloadcms/payload/discussions/1961

**Comment by wkentdag** [comment]
I stumbled across this issue while migrating a set of Wordpress uploads to payload. The WP URL pattern `/yyyy/mm/dd/<filename>` gets flattened to `/yyyymmdd<filename>`, which breaks any existing asset URL. The workarounds mentioned in #1153 don't help in this case since the `/` chars ultimately get stripped [here](https://github.com/payloadcms/payload/blob/main/packages/payload/src/uploads/getSafeFilename.ts#L10).

I wound up using some Next js [middleware hackery](https://discord.com/channels/967097582721572934/1231361822146170900/1231387409023897730) to mostly sidestep this limitation. But IMO a cleaner solution would be to expose a `sanitizeFilenameOptions` config option where consumers can pass their own [options](https://github.com/parshap/node-sanitize-filename?tab=readme-ov-file#api) directly to `sanitize-filename`. How does that sound?

**Reply by mrobst** [reply]
If I understand the santizeFilenameOptions correctly you can pass in a restricted character and define what to get back in that instance.... so ("/" -> "/")  would probably solve the use case for cloud storage with virtual folders. The challenge is how to get the options exposed to payload.config.ts !

**Comment by tdefilippis** [comment]
Has this been solved?

**Reply by mrobst** [reply]
Not as far as I know

**Comment by anthonymzupancic** [comment]
Adding here to hopefully bump this again. I'm working with the Google Cloud Storage adapter and was able to successfully add a folder structure/file path through the use of the `beforeOperation` hook and setting the `prefix` for the file write. I'm trying to use the subdomain as the folder for the multi-tenant plugin use-case.

The issue I'm running into is being able to read/use the file. I know the admin thumb needs to be redirected, but I cannot for the life of me get the folder'd files to be found. I was able to append the prefix on read, but that responds with a 404 because it's not altering the file path to be read, but the API route to read from which is the 404.

**Reply by mrobst** [reply]
I've got my scenario all working but its it two parts. 

The file upload assigns a prefix e.g. the subdomain. In storage the files have a subdomain prefix and collisions work as they should (you can have the same file in different subdomains). Note you need to specify the filename compound index and make sure that any filename index on the collection is dropped for this to work https://payloadcms.com/docs/upload/overview#collection-upload-options

For the read side you can't use the prefix so you need to assign the same subdomain to the document in the collection. In my scenario I'm using a hook to assign the subdomain as a relationship to a collection of subdomains. Using normal access control you can control which subdomains (field in collection) can be read/written etc. Using a prefix hook you can keep the storage separate. As long as the logic for the two hooks aligns it all works great. 

This, plus the changes recently to the storage adapters where you can have different buckets/containers for different collections are a huge step forward for storage management in PayloadCMS :-) :-)

**Comment by mrobst** [comment]
This issue is at least partly solved now  #14566 
Also the recent change to use a map for storage clients so that each collection can have its own container/bucket is great. 

Testing to see if it all hangs together, will report back
